logstreams/src/main/java/io/camunda/zeebe/logstreams/impl/log/Sequencer.java

```diff
@@ -0,0 +1,196 @@
+   * @param batchSize the potential batch Size (in bytes) we want to check
+   * @return True if the serialized batch would fit within {@link Sequencer#maxFragmentSize}, false
+   *     otherwise.
+   */
+  @Override
+  public boolean canWriteEvents(final int eventCount, final int batchSize) {
+    final int framedMessageLength =
+        batchSize
+            + eventCount * (DataFrameDescriptor.HEADER_LENGTH + FRAME_ALIGNMENT)
+            + FRAME_ALIGNMENT;
+    return framedMessageLength <= maxFragmentSize;
+  }
+
+  /**
+   * @param appendEntry the entry to write
+   * @param sourcePosition a back-pointer to the record whose processing created this entry
+   * @return -1 if write was rejected, the position of the entry if write was successful.
+   */
+  @Override
+  public long tryWrite(final LogAppendEntry appendEntry, final long sourcePosition) {
+    if (isClosed) {
+      LOG.warn("Rejecting write of {}, sequencer is closed", appendEntry);
+      return -1;
+    }
+    final boolean isEnqueued;
+    final long currentPosition;
+
+    lock.lock();
+    try {
+      currentPosition = position;
+      isEnqueued =
+          queue.offer(new SequencedBatch(currentPosition, sourcePosition, List.of(appendEntry)));
+      if (isEnqueued) {
+        position = currentPosition + 1;
+      }
+    } finally {
+      lock.unlock();
+    }
+
+    if (consumer != null) {
+      consumer.signal();
+    }
+    metrics.observeBatchSize(1);
+    metrics.setQueueSize(queue.size());
+    if (isEnqueued) {
+      return currentPosition;
+    } else {
+      LOG.trace("Rejecting write of {}, sequencer queue is full", appendEntry);
+      return -1;
+    }
+  }
+
+  /**
+   * @param appendEntries a set of entries to append; these will be appended in the order in which
+   *     the collection is iterated.
+   * @param sourcePosition a back-pointer to the record whose processing created these entries
+   * @return -1 if write was rejected, 0 if batch was empty, the highest position of the batch if
+   *     write was successful.
+   */
+  @Override
+  public long tryWrite(
+      final Iterable<? extends LogAppendEntry> appendEntries, final long sourcePosition) {
+    if (isClosed) {
+      LOG.warn("Rejecting write of {}, sequencer is closed", appendEntries);
+      return -1;
+    }
+
+    final var entries = new ArrayList<LogAppendEntry>(16);
+    for (final var entry : appendEntries) {
+      entries.add(entry);
+    }
+    final var batchSize = entries.size();
+    if (batchSize == 0) {
+      return 0;
+    }
+
+    final long currentPosition;
+    final boolean isEnqueued;
+    lock.lock();
+    try {
+      currentPosition = position;
+      isEnqueued = queue.offer(new SequencedBatch(currentPosition, sourcePosition, entries));
+      if (isEnqueued) {
+        position = currentPosition + batchSize;
+      }
+    } finally {
+      lock.unlock();
+    }
+
+    if (consumer != null) {
+      consumer.signal();
```

Comment:
:x: Same as above - or am I missing something? :thinking: Is it because if the queue is full we want to ensure the consumer is triggered to fetch from it?

Reply:
Yes, that's exactly it.